src/analyses/relationAnalysis.ml

```diff
@@ -0,0 +1,692 @@
+    match lv with
+    | (Var v, NoOffset) when RD.Tracked.varinfo_tracked v ->
+      if not v.vglob && not (ThreadEscape.has_escaped ask v) then (
+        if ask.f (Queries.IsMultiple v) then
+          {st with rel = RD.join (f st v) st.rel}
+        else
+          {st with rel = f st v}
+      )
+      else (
+        let v_out = Goblintutil.create_var @@ makeVarinfo false (v.vname ^ "#out") v.vtype in (* temporary local g#out for global g *)
+        v_out.vattr <- v.vattr; (*copy the attributes because the tracking may depend on them. Otherwise an assertion fails *)
+        let st = {st with rel = RD.add_vars st.rel [RV.local v_out]} in (* add temporary g#out *)
+        let st' = {st with rel = f st v_out} in (* g#out = e; *)
+        if M.tracing then M.trace "relation" "write_global %a %a\n" d_varinfo v d_varinfo v_out;
+        let st' = write_global ask getg sideg st' v v_out in (* g = g#out; *)
+        let rel'' = RD.remove_vars st'.rel [RV.local v_out] in (* remove temporary g#out *)
+        {st' with rel = rel''}
+      )
+    | (Mem v, NoOffset) ->
+      (let r = ask.f (Queries.MayPointTo v) in
+       match r with
+       | `Top ->
+         st
+       | `Lifted s ->
+         let lvals = Queries.LS.elements r in
+         let ass' = List.map (fun lv -> assign_to_global_wrapper ask getg sideg st (Lval.CilLval.to_lval lv) f) lvals in
+         List.fold_right D.join ass' (D.bot ())
+      )
+    (* Ignoring all other assigns *)
+    | _ ->
+      st
+
+
+  (** An extended overflow handling inside relationAnalysis for expression assignments when overflows are assumed to occur.
+      Since affine equalities can only keep track of integer bounds of expressions evaluating to definite constants, we now query the integer bounds information for expressions from other analysis.
+      If an analysis returns bounds that are unequal to min and max of ikind , we can exclude the possibility that an overflow occurs and the abstract effect of the expression assignment can be used, i.e. we do not have to set the variable's value to top. *)
+
+  let no_overflow (ask: Queries.ask) exp =
+    match Cilfacade.get_ikind_exp exp with
+    | exception Invalid_argument _ -> false (* TODO: why this? *)
+    | exception Cilfacade.TypeOfError _ -> false
+    | ik ->
+      if IntDomain.should_wrap ik then
+        false (* TODO: why this? *)
+      else if IntDomain.should_ignore_overflow ik then
+        true
+      else
+        let eval_int exp = Queries.ID.to_bool (ask.f (EvalInt exp)) in
+        (* TODO: duplicates assert_type_bounds *)
+        let type_min, type_max = IntDomain.Size.range ik in
+        (* TODO: why not Le? *)
+        let min = eval_int (BinOp (Lt, kintegerCilint ik type_min, exp, intType)) in
+        let max = eval_int (BinOp (Lt, exp, kintegerCilint ik type_max, intType)) in
+        match min, max with (* TODO: don't need both? *)
+        | Some true, _
+        | _, Some true -> true
+        | _, _ -> false
```

Comment:
Am I the only one to whom this no overflow check doesn't make sense? It checks both bounds with strict inequality, although equaling a type bound is also overflow-free, and only one of the two bounds needs to fit the type. With wraparound, it doesn't even bother doing the bounds checks although they might actually still fit.

Reply:
For both exceptions above, `false` is the safe default. Otherwise this is a bit of a hack that we did because we were too lazy for a proper solution.

In cases where we are ignoring overflows, we just return true.

The values I get from `eval_int` have already been handling for overflow. The thing we abuse here is that if an overflow occurred and we did go to top, the upper AND lower bound will be top, if either of these is a non-top value, now overflow can have happened.

This come at the price that we sometimes spuriously assume overflows where it would not be necessary as the value simply is top without any overflows happening.

This logic does however not apply for unsigned types as they can have overflowed without the resulting value being top.

This is all a bit hacky, and maybe one should install proper `MayOverflow` queries later, but I think one can leave this for a follow-up PR.

Reply:
Yes, it should!